drivers: serial: drop uart_device_config

gmarull opened the pull request:
This PR removes `struct uart_device_config` from drivers using it, and does some cleanups thanks to this removal. The main points to remove this structure:

1. It is not used by all drivers (in fact, I think it's a minority using it)
2. It pretends to be generic, but it is not. For example, the field that stores the register base address is not suitable for most drivers.
3. No other subsystems are using this kind of inheritance mechanism (I think).
4. Some fields are not always used, wasting extra ROM space (e.g. sys_clk_freq).

Fixes #42130

gmarull committed:
drivers: serial: stm32: store USART instance

Instead of using "generic" uart_device_config fields, store the right
pointer to avoid unnecessary casts. This change makes code simpler and
more idiomatic.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: stm32: store IRQ config function in a single place

When CONFIG_PM=y, the IRQ function also needs to be stored, something
the "generic" uart_device_config cannot support. This uses the config
irq_config_func field to store the function in all situations, thus
removing unnecessary logic and finally dropping uart_device_config.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

erwango reviewed (dismissed):
Thanks for the cleanup!

gmarull committed:
drivers: serial: altera_jtag: remove unused config

The driver configuration was not used, so just remove it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: cc13xx_cc26xx: drop usage of uart_device_config

Create a driver specific configuration structure, containing the
required fields only.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: cc32xx: drop usage of uart_device_config

Create a driver specific configuration structure, containing the
required fields only.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: cmsdk_apb: drop usage of uart_device_config

Create a driver specific configuration structure, containing the
required fields only. Since the config struct can now store a pointer to
the UART structure, casts from (uint8_t *) to (struct uart_cmsdk_apb*)
are no longer needed. UART_STRUCT has also been dropped in favor of
using the config pointer directly now that it is possible.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: msp432p4xx: remove unused sys_clk_freq

The variable is not used by the driver, so there is no need to
initialize it.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: msp432p4xx: drop usage of uart_device_config

Create a driver specific configuration structure, containing the
required fields only.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: npcx: drop usage of uart_device_config

Create a driver specific configuration structure, containing the
required fields only. Since the config struct can now store a pointer to
the UART structure, casts from address to (struct uart_reg *) are no
longer needed. HAL_INSTANCE has also been dropped in favor of using the
uart field from the config struct now that it is possible.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: numicro: drop usage of uart_device_config

Create a driver specific configuration structure, containing the
required fields only. Since the config struct can now store a pointer to
the UART structure, casts from address to (UART_T *) are no longer
needed. UART_STRUCT has also been dropped in favor of using the config
pointer directly now that it is possible.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: pl011: drop usage of uart_device_config

Create a driver specific configuration structure, containing the
required fields only. Since the config struct can now store a pointer to
the UART structure, casts from address to (struct pl011_regs *) are no
longer needed. PL011_REGS has also been dropped in favor of using the
config pointer directly now that it is possible.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: stellaris: drop usage of uart_device_config

Create a driver specific configuration structure, containing the
required fields only. Since the config struct can now store a pointer to
the UART structure, casts from address to (struct _uart*) are no longer
needed. UART_STRUCT has also been dropped in favor of using the config
pointer directly now that it is possible.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: stellaris: remove unused definitions

The register definitions are not used by the driver (struct _uart is
used instead).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: xlnx_ps: drop usage of uart_device_config

Create a driver specific configuration structure, containing the
required fields only.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: xmc4xxx: drop usage of uart_device_config

Create a driver specific configuration structure, containing the
required fields only.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: uart: remove uart_device_config

This structure is no longer used by drivers. Instead, each driver is
responsible to define the required fields and the types that are needed.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

ChiHuaL commented on drivers/serial/uart_npcx.c:
```diff
@@ -102,42 +101,43 @@ static int uart_set_npcx_baud_rate(struct uart_reg *const inst, int baud_rate, i
 static int uart_npcx_tx_fifo_ready(const struct device *dev)
 {
-	struct uart_reg *const inst = HAL_INSTANCE(dev);
+	const struct uart_npcx_config *const config = dev->config;
```
How about:
struct uart_reg *const reg_base =dev->config->uart;
So we can just access the UART register like:
GET_FIELD(reg_base->UFTSTS, NPCX_UFTSTS_TEMPTY_LVL)

ChiHuaL reviewed (commented):
1 nit for UART register access.
